### Import libraries

In [1]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt

### Import data

In [2]:
path = 'https://raw.githubusercontent.com/dan1dr/microdosing/main/data/dlongcodes.csv'
data = pd.read_csv(path, sep=',', index_col=False)
pd.options.display.max_columns = None

In [3]:
data.head(10)

ResponseId         StartDate           EndDate      RecordedDate  \
0  R_0HSeYpblz5LCLpD  19/09/2017 18:27  19/09/2017 18:51  19/09/2017 18:51   
1  R_0HSeYpblz5LCLpD  19/09/2017 18:27  19/09/2017 18:51  19/09/2017 18:51   
2  R_0HSeYpblz5LCLpD  19/09/2017 18:27  19/09/2017 18:51  19/09/2017 18:51   
3  R_0HSeYpblz5LCLpD  19/09/2017 18:27  19/09/2017 18:51  19/09/2017 18:51   
4  R_0HSeYpblz5LCLpD  19/09/2017 18:27  19/09/2017 18:51  19/09/2017 18:51   
5  R_0HSeYpblz5LCLpD  19/09/2017 18:27  19/09/2017 18:51  19/09/2017 18:51   
6  R_0MpphsGkrwSMhuF  10/10/2017 21:06  10/10/2017 21:10  24/10/2017 21:10   
7  R_0MpphsGkrwSMhuF  10/10/2017 21:06  10/10/2017 21:10  24/10/2017 21:10   
8  R_0MpphsGkrwSMhuF  10/10/2017 21:06  10/10/2017 21:10  24/10/2017 21:10   
9  R_0MpphsGkrwSMhuF  10/10/2017 21:06  10/10/2017 21:10  24/10/2017 21:10   

   MoodBoard_1  MoodBoard_2  MoodBoard_3  MoodBoard_4  MoodBoard_5  \
0            0            1            1            1            1   
1            0            1            1            1            1   
2            0            1            1            1            1   
3            0            1            1            1            1   
4            0            1            1            1            1   
5            0            1            1            1            1   
6            0            0            0            1            0   
7            0            0            0            1            0   
8            0            0            0            1            0   
9            0            0            0            1            0   

   MoodBoard_6  MoodBoard_7  MoodBoard_8  MoodBoard_9  MoodBoard_10  \
0            1            1            1            0             0   
1            1            1            1            0             0   
2            1            1            1            0             0   
3            1            1            1            0             0   
4            1            1            1            0             0   
5            1            1            1            0             0   
6            0            0            0            0             0   
7            0            0            0            0             0   
8            0            0            0            0             0   
9            0            0            0            0             0   

   MoodBoard_11  MoodBoard_12  MoodBoard_13  MoodBoard_14  MoodBoard_15  \
0             1             0             1             0             0   
1             1             0             1             0             0   
2             1             0             1             0             0   
3             1             0             1             0             0   
4             1             0             1             0             0   
5             1             0             1             0             0   
6             0             0             0             0             0   
7             0             0             0             0             0   
8             0             0             0             0             0   
9             0             0             0             0             0   

   MoodBoard_16  MoodBoard_17  MoodBoard_18  MoodBoard_19  MoodBoard_20  \
0             0             1             1             0             1   
1             0             1             1             0             1   
2             0             1             1             0             1   
3             0             1             1             0             1   
4             0             1             1             0             1   
5             0             1             1             0             1   
6             0             1             0             0             0   
7             0             1             0             0             0   
8             0             1             0             0             0   
9             0       

In [4]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1668 entries, 0 to 1667
Columns: 242 entries, ResponseId to ConCategory
dtypes: float64(138), int64(64), object(40)
memory usage: 3.1+ MB


In [5]:
data.shape

(1668, 242)

In [6]:
data.columns

Index(['ResponseId', 'StartDate', 'EndDate', 'RecordedDate', 'MoodBoard_1',
       'MoodBoard_2', 'MoodBoard_3', 'MoodBoard_4', 'MoodBoard_5',
       'MoodBoard_6',
       ...
       'ProRating', 'ResponseId.2', 'ConIndex', 'ConCode', 'Rating.C',
       'ConRating', 'ProConcept', 'ProCategory', 'ConConcept', 'ConCategory'],
      dtype='object', length=242)

## NAN's and duplicates

Since data comes from a survey conducted on the Internet, it may be sometimes be struggling to normalize and clean the format. Thankfully, this dataset was previously used and pre-cleaned by different authors https://osf.io/g5cwy/

However, the format and questionnaire performed is quite complex and we will need to make some adjustments.

Firstly, we need to assess the NANs, duplicates and other elements

In [7]:
def assess_NA(data):
    """
    Returns a pandas dataframe denoting the total number of NA
    values and the percentage of NA values in each column.
    The column names are noted on the index.
    """
    # pandas series denoting features and the sum of their null values
    null_sum = data.isnull().sum()# instantiate columns for missing data
    total = null_sum.sort_values(ascending=False)
    percent = ( ((null_sum / len(data.index))*100).round(2) ).sort_values(ascending=False)
    
    # concatenate along the columns to create the complete dataframe
    df_NA = pd.concat([total, percent], axis=1, sort=True, keys=['Number of NA', 'Percent NA'])
    
    # drop rows that don't have any missing data; omit if you want to keep all rows
    df_NA = df_NA[ (df_NA.T != 0).any() ]
    
    return df_NA
    
    
    

In [8]:
assess_NA(data)

Number of NA  Percent NA
BFI.2_63                            474       28.42
BFI.2_64                            474       28.42
BFI.2_65                            480       28.78
BFI.2_66                            474       28.42
BFI.2_67                            474       28.42
...                                 ...         ...
RedTobacco                           66        3.96
Reddit_5_TEXT...Topics             1308       78.42
Religion_12_TEXT...Topics          1602       96.04
SES                                  42        2.52
Substance_8_TEXT...Topics          1668      100.00

[160 rows x 2 columns]

Data is relatively "clean" and we will use all the rows given, most of NAs come from fields where the response ask for an alternative text in case of not being listed in the others before, so it is not relevant.

In [9]:
data['ResponseId'].nunique()

278

As we see, the data consists on a set of 278 participants identified by ``ResponseId``. The number of rows is 1668 because the table is given in a "long" format, where the pivot variable is the identity and repeats the same pattern every 6 rows (1668/6 = 278).

This will be important when dealing beyond the 'ProIndex' column, until that point we can dismiss this fact and work only with the unique values in order to acommodate the dataset.

Before splitting the dataset into two different parts, we will rename some columns in order to facilitate future tasks.

In [15]:
data = data.rename(columns={"MoodBoard_1": "Angry", "MoodBoard_2": "Distressed",
                    "MoodBoard_3": "Afraid", "MoodBoard_4": "Frustrated",
                    "MoodBoard_5": "Nervous", "MoodBoard_6": "Alarmed",
                    "MoodBoard_7": "Tense", "MoodBoard_8": "Annoyed",
                    "MoodBoard_9": "Astonished", "MoodBoard_10": "Excited",
                    "MoodBoard_11": "Alert", "MoodBoard_12": "Delighted",
                    "MoodBoard_13": "Amused", "MoodBoard_14": "Glad",
                    "MoodBoard_15": "Happy", "MoodBoard_16": "Pleased",
                    "MoodBoard_17": "Sad", "MoodBoard_18": "Miserable",
                    "MoodBoard_19": "Depressed", "MoodBoard_20": "Droopy",
                    "MoodBoard_21": "Gloomy", "MoodBoard_22": "Fatigued",
                    "MoodBoard_23": "Bored", "MoodBoard_24": "Tired",
                    "MoodBoard_25": "Serene", "MoodBoard_26": "Content",
                    "MoodBoard_27": "Calm", "MoodBoard_28": "At ease",
                    "MoodBoard_29": "Satisfied", "MoodBoard_30": "Relaxed",
                    "MoodBoard_31": "Sleepy", "MoodBoard_32": "Quiet"}, errors="raise")

In [17]:
b = data.drop_duplicates(subset = 'ResponseId')
b.shape

(278, 242)

In [18]:
b.index

Int64Index([   0,    6,   12,   18,   24,   30,   36,   42,   48,   54,
            ...
            1608, 1614, 1620, 1626, 1632, 1638, 1644, 1650, 1656, 1662],
           dtype='int64', length=278)

In [19]:
#Rearrange the index from 0
b.index = range(1,279)
b.head()

ResponseId         StartDate           EndDate      RecordedDate  \
1  R_0HSeYpblz5LCLpD  19/09/2017 18:27  19/09/2017 18:51  19/09/2017 18:51   
2  R_0MpphsGkrwSMhuF  10/10/2017 21:06  10/10/2017 21:10  24/10/2017 21:10   
3  R_0N8EpmmQ5b79cxr  18/10/2017 20:29  18/10/2017 23:51  18/10/2017 23:51   
4  R_0Nj0uZDZlPP0IRX  18/09/2017 11:40  18/09/2017 11:58  18/09/2017 11:58   
5  R_0OZWDFj2krlpms9  22/09/2017 18:03  23/09/2017 06:47  07/10/2017 06:47   

   Angry  Distressed  Afraid  Frustrated  Nervous  Alarmed  Tense  Annoyed  \
1      0           1       1           1        1        1      1        1   
2      0           0       0           1        0        0      0        0   
3      0           0       0           0        0        0      0        0   
4      0           0       0           0        0        0      0        0   
5      0           1       0           0        1        0      0        0   

   Astonished  Excited  Alert  Delighted  Amused  Glad  Happy  Pleased  Sad  \
1           0        0      1          0       1     0      0        0    1   
2           0        0      0          0       0     0      0        0    1   
3           0        0      0          0       0     0      1        0    0   
4           0        0      0          0       0     0      0        0    0   
5           0        1      1          0       1     0      1        1    0   

   Miserable  Depressed  Droopy  Gloomy  Fatigued  Bored  Tired  Serene  \
1          1          0       1       1         0      0      1       0   
2          0          0       0       0         1      1      1       0   
3          0          0       1       0         1      0      1       0   
4          0          0       1       0         1      0      0       0   
5          0          0       1       0         0      0      1       1   

   Content  Calm  At ease  Satisfied  Relaxed  Sleepy  Quiet  Current  Reddit  \
1        0     0        0          0        0       1      0        1       5   
2        1     1        0          0        1       0      1        1       1   
3        1     0        0          0        1       1      0        2       2   
4        0     0        0          0        0       0      0        1       1   
5        1     0        1          1        1       1      0        2       5   

   Country  Age  Gender  Orientation  Ethnicity Ethnicity_11_TEXT  Religion  \
1      187   21    -1.0         -1.0          1               -99         1   
2      187   29    -1.0         -1.0          6               -99         1   
3      187   20    -1.0         -1.0         13               -99         1   
4      128   31    -1.0         -1.0          1               -99         1   
5       -1   27    -1.0         -1.0          1               -99        14   

  Religion_12_TEXT  Religiosity  Spirituality  Education  SES  AgeMD  \
1              -99            2            20        6.0  1.0     20   
2              -99           14            19        7.0  2.0     28   
3              -99            0             5        3.0  2.0     19   
4              -99            0            28        6.0  1.0     32   
5              -99            2            68        6.0  2.0     26   

   Substance_1  Substance_2  Substance_8  Dose_1  Dose_4  Dose_2  Dose_3  \
1            1          -99          -99     -99       1     -99     -99   
2            1            1          -99     -99       1     -99     -99   
3            1          -99          -99       1       1     -99     -99   
4            1          -99          -99     -99       1     -99     -99   
5          -99            1          -99     -99     -99       1     -99   

   Dose_7  Dose_8  Spacing  DoseNumber  Money Money_1_TEXT Money_2_TEXT  \
1     -99     -99        3        25.0      1            5          -99   
2     -99     -99        1        26.0      1           20          -99   
3     -99     -99        5        25.0      1           10          -

Beyond the 'ProIndex' columns, we find the reason of the long format. Exploring the original data https://osf.io/dwtk5/ we find that there are represented the following values: the first three correspond to "pro_text_X_topics", and the last three to "pro_X_code".

The first 3 rows are almost identical to the following 3, just differ when sometimes one is NA or the topics is different from the category. We will select the category as it is a more general characteristic. Hence, we will only use the last 3 rows per each ``ResponseId``

## From wide to long

In [20]:
long_info = data.loc[:, 'ProIndex':]
long_info.head(12)

ProIndex               ProCode       ResponseId.1  Rating.P  ProRating  \
0          1         Introspection  R_0HSeYpblz5LCLpD         1       85.0   
1          2  Perspective Shifting  R_0HSeYpblz5LCLpD         2       90.0   
2          3                   NaN  R_0HSeYpblz5LCLpD         3       55.0   
3          4         Introspection  R_0HSeYpblz5LCLpD         4       85.0   
4          5  Perspective Shifting  R_0HSeYpblz5LCLpD         5       90.0   
5          6             Self-Care  R_0HSeYpblz5LCLpD         6       55.0   
6          1            Creativity  R_0MpphsGkrwSMhuF         1       70.0   
7          2   Improved Motivation  R_0MpphsGkrwSMhuF         2       52.0   
8          3         Improved Mood  R_0MpphsGkrwSMhuF         3       45.0   
9          4            Creativity  R_0MpphsGkrwSMhuF         4       70.0   
10         5   Improved Motivation  R_0MpphsGkrwSMhuF         5       52.0   
11         6         Improved Mood  R_0MpphsGkrwSMhuF         6       45.0   

         ResponseId.2  ConIndex             ConCode  Rating.C  ConRating  \
0   R_0HSeYpblz5LCLpD         1  Sleep Disturbances         1       25.0   
1   R_0HSeYpblz5LCLpD         2          Illegality         2        4.0   
2   R_0HSeYpblz5LCLpD         3              Stigma         3       20.0   
3   R_0HSeYpblz5LCLpD         4  Sleep Disturbances         4       25.0   
4   R_0HSeYpblz5LCLpD         5          Illegality         5        4.0   
5   R_0HSeYpblz5LCLpD         6              Stigma         6       20.0   
6   R_0MpphsGkrwSMhuF         1  Sleep Disturbances         1       35.0   
7   R_0MpphsGkrwSMhuF         2                 NaN         2        NaN   
8   R_0MpphsGkrwSMhuF         3                 NaN         3        NaN   
9   R_0MpphsGkrwSMhuF         4  Sleep Disturbances         4       35.0   
10  R_0MpphsGkrwSMhuF         5                 NaN         5        NaN   
11  R_0MpphsGkrwSMhuF         6                 NaN         6        NaN   

                 ProConcept    ProCategory                  ConConcept  \
0                 Self-Care  Self-Efficacy  Physiological Interference   
1   Meta-Creative Processes     Creativity                  Illegality   
2                       NaN            NaN                      Stigma   
3                 Self-Care  Self-Efficacy  Physiological Interference   
4   Meta-Creative Processes     Creativity                  Illegality   
5                 Self-Care  Self-Efficacy                      Stigma   
6                Creativity     Creativity  Physiological Interference   
7             Self-Efficacy  Self-Efficacy                         NaN   
8             Improved Mood  Improved Mood                         NaN   
9                Creativity     Creativity  Physiological Interference   
10            Self-Efficacy  Self-Efficacy                         NaN   
11            Improved Mood  Improved Mood                         NaN   

                 ConCategory  
0   Physiological Discomfort  
1                 Illegality  
2                 Illegality  
3   Physiological Discomfort  
4                 Illegality  
5                 Illegality  
6   Physiological Discomfort  
7                        NaN  
8                        NaN  
9   Physiological Discomfort  
10                       NaN  
11                       NaN

In [538]:
x =  data.loc[:, 'ProIndex':]
x = x.drop(columns=['ResponseId.2','Rating.P'])
x.index = range(1,len(x)+1)

Now let's rearrange the columns to facilitate the comparison between pros and cons.

In [422]:
x.columns

Index(['ProIndex', 'ProCode', 'ResponseId.1', 'ProRating', 'ConIndex',
       'ConCode', 'Rating.C', 'ConRating', 'ProConcept', 'ProCategory',
       'ConConcept', 'ConCategory'],
      dtype='object')

In [400]:
def return_column_order(dataframe, order, rest=False, front=True):
    '''Takes a dataframe and reorder its columns. If "rest" is 
       True, it also takes into account the columns not listed, then 
       returns dataframe with order as first columns if "front" is 
       True, and order as last columns if "front" is False.
    '''
    cols = order[:]
    if rest:
        for x in dataframe.columns:
            if x not in cols:
                if front: #insert in the front in the list
                    cols.append(x)
                else: #insert to be in the last
                    cols.insert(0, x)
    return dataframe[cols]

In [539]:
x = return_column_order(x, ['ResponseId.1', 'ProCode', 'ProCategory',
                        'ProRating', 'ConCode', 'ConCategory',
                        'ConRating'])

Now we want to pivot from the "RespondeId.1" and transform it to a wide data. We will only use the last 3 rows per respondent as discussed before. So previous to pivoting let's select 3 rows per user and later create ProCodeX, ProCategory and ProRatingX for i from 1 to 3. Same proccess with Cons.

In [402]:
x

ResponseId.1               ProCode         ProCategory  ProRating  \
1     R_0HSeYpblz5LCLpD         Introspection       Self-Efficacy       85.0   
2     R_0HSeYpblz5LCLpD  Perspective Shifting          Creativity       90.0   
3     R_0HSeYpblz5LCLpD                   NaN                 NaN       55.0   
4     R_0HSeYpblz5LCLpD         Introspection       Self-Efficacy       85.0   
5     R_0HSeYpblz5LCLpD  Perspective Shifting          Creativity       90.0   
...                 ...                   ...                 ...        ...   
1664  R_ZfYg5JxACJX7szf    Clarity of Thought  Cognitive Benefits       79.0   
1665  R_ZfYg5JxACJX7szf         Improved Mood       Improved Mood      100.0   
1666  R_ZfYg5JxACJX7szf           Sociability     Social Benefits       49.0   
1667  R_ZfYg5JxACJX7szf    Clarity of Thought  Cognitive Benefits       79.0   
1668  R_ZfYg5JxACJX7szf         Improved Mood       Improved Mood      100.0   

                          ConCode                ConCategory  ConRating  
1              Sleep Disturbances   Physiological Discomfort       25.0  
2                      Illegality                 Illegality        4.0  
3                          Stigma                 Illegality       20.0  
4              Sleep Disturbances   Physiological Discomfort       25.0  
5                      Illegality                 Illegality        4.0  
...                           ...                        ...        ...  
1664                          NaN                        NaN       59.0  
1665                          NaN                        NaN        9.0  
1666                         None  Other Perceived Drawbacks        7.0  
1667  Unknown Risk-Effect Profile  Other Perceived Drawbacks       59.0  
1668                          NaN                        NaN        9.0  

[1668 rows x 7 columns]

In [540]:
type(len(x)/x['ResponseId.1'].nunique())

float

In [542]:
list = []
    #step in range must be converted into a integer
for i in range(3, len(x), int(len(x)/x['ResponseId.1'].nunique())):
    list.append(i)
    list.append(i+1)
    list.append(i+2)

print(list)

[3, 4, 5, 9, 10, 11, 15, 16, 17, 21, 22, 23, 27, 28, 29, 33, 34, 35, 39, 40, 41, 45, 46, 47, 51, 52, 53, 57, 58, 59, 63, 64, 65, 69, 70, 71, 75, 76, 77, 81, 82, 83, 87, 88, 89, 93, 94, 95, 99, 100, 101, 105, 106, 107, 111, 112, 113, 117, 118, 119, 123, 124, 125, 129, 130, 131, 135, 136, 137, 141, 142, 143, 147, 148, 149, 153, 154, 155, 159, 160, 161, 165, 166, 167, 171, 172, 173, 177, 178, 179, 183, 184, 185, 189, 190, 191, 195, 196, 197, 201, 202, 203, 207, 208, 209, 213, 214, 215, 219, 220, 221, 225, 226, 227, 231, 232, 233, 237, 238, 239, 243, 244, 245, 249, 250, 251, 255, 256, 257, 261, 262, 263, 267, 268, 269, 273, 274, 275, 279, 280, 281, 285, 286, 287, 291, 292, 293, 297, 298, 299, 303, 304, 305, 309, 310, 311, 315, 316, 317, 321, 322, 323, 327, 328, 329, 333, 334, 335, 339, 340, 341, 345, 346, 347, 351, 352, 353, 357, 358, 359, 363, 364, 365, 369, 370, 371, 375, 376, 377, 381, 382, 383, 387, 388, 389, 393, 394, 395, 399, 400, 401, 405, 406, 407, 411, 412, 413, 417, 418, 419, 42

In [450]:
def select_last3(dataframe):
    list = []
    #step in range must be converted into a integer
    for i in range(3, len(x), int(len(x)/x['ResponseId.1'].nunique())):
        list.append(i)
        list.append(i+1)
        list.append(i+2)
    return dataframe.iloc[list]

In [543]:
x = select_last3(x)
x

ResponseId.1               ProCode         ProCategory  ProRating  \
4     R_0HSeYpblz5LCLpD         Introspection       Self-Efficacy       85.0   
5     R_0HSeYpblz5LCLpD  Perspective Shifting          Creativity       90.0   
6     R_0HSeYpblz5LCLpD             Self-Care       Self-Efficacy       55.0   
10    R_0MpphsGkrwSMhuF            Creativity          Creativity       70.0   
11    R_0MpphsGkrwSMhuF   Improved Motivation       Self-Efficacy       52.0   
...                 ...                   ...                 ...        ...   
1661  R_yUcpPqMIp5DPBSN       Improved Energy     Improved Energy       80.0   
1662  R_yUcpPqMIp5DPBSN        Improved Focus      Improved Focus      100.0   
1666  R_ZfYg5JxACJX7szf           Sociability     Social Benefits       49.0   
1667  R_ZfYg5JxACJX7szf    Clarity of Thought  Cognitive Benefits       79.0   
1668  R_ZfYg5JxACJX7szf         Improved Mood       Improved Mood      100.0   

                          ConCode                ConCategory  ConRating  
4              Sleep Disturbances   Physiological Discomfort       25.0  
5                      Illegality                 Illegality        4.0  
6                          Stigma                 Illegality       20.0  
10             Sleep Disturbances   Physiological Discomfort       35.0  
11                            NaN                        NaN        NaN  
...                           ...                        ...        ...  
1661                  Preparation  Other Perceived Drawbacks        5.0  
1662                         Misc  Other Perceived Drawbacks        5.0  
1666                         None  Other Perceived Drawbacks        7.0  
1667  Unknown Risk-Effect Profile  Other Perceived Drawbacks       59.0  
1668                          NaN                        NaN        9.0  

[834 rows x 7 columns]

Another strategy could have been using the 'ProIndex' column previously deleted and selecting the values from 4 to 6 per ``ResponseId.1``.

Now we have all we need to pivot the data

In [559]:
df = pd.DataFrame(index=range(1,279),columns= ['ProCode1', 'ProCategory1', 'ProRating1',
                 'ConCode1', 'ConCategory1', 'ConRating1',
                'ProCode2', 'ProCategory2', 'ProRating2',
                 'ConCode2', 'ConCategory2', 'ConRating2',
                'ProCode3', 'ProCategory3', 'ProRating3',
                 'ConCode3', 'ConCategory1', 'ConRating3'])
df

ProCode1 ProCategory1 ProRating1 ConCode1 ConCategory1 ConRating1  \
1        NaN          NaN        NaN      NaN          NaN        NaN   
2        NaN          NaN        NaN      NaN          NaN        NaN   
3        NaN          NaN        NaN      NaN          NaN        NaN   
4        NaN          NaN        NaN      NaN          NaN        NaN   
5        NaN          NaN        NaN      NaN          NaN        NaN   
..       ...          ...        ...      ...          ...        ...   
274      NaN          NaN        NaN      NaN          NaN        NaN   
275      NaN          NaN        NaN      NaN          NaN        NaN   
276      NaN          NaN        NaN      NaN          NaN        NaN   
277      NaN          NaN        NaN      NaN          NaN        NaN   
278      NaN          NaN        NaN      NaN          NaN        NaN   

    ProCode2 ProCategory2 ProRating2 ConCode2 ConCategory2 ConRating2  \
1        NaN          NaN        NaN      NaN          NaN        NaN   
2        NaN          NaN        NaN      NaN          NaN        NaN   
3        NaN          NaN        NaN      NaN          NaN        NaN   
4        NaN          NaN        NaN      NaN          NaN        NaN   
5        NaN          NaN        NaN      NaN          NaN        NaN   
..       ...          ...        ...      ...          ...        ...   
274      NaN          NaN        NaN      NaN          NaN        NaN   
275      NaN          NaN        NaN      NaN          NaN        NaN   
276      NaN          NaN        NaN      NaN          NaN        NaN   
277      NaN          NaN        NaN      NaN          NaN        NaN   
278      NaN          NaN        NaN      NaN          NaN        NaN   

    ProCode3 ProCategory3 ProRating3 ConCode3 ConCategory1 ConRating3  
1        NaN          NaN        NaN      NaN          NaN        NaN  
2        NaN          NaN        NaN      NaN          NaN        NaN  
3        NaN          NaN        NaN      NaN          NaN        NaN  
4        NaN          NaN        NaN      NaN          NaN        NaN  
5        NaN          NaN        NaN      NaN          NaN        NaN  
..       ...          ...        ...      ...          ...        ...  
274      NaN          NaN        NaN      NaN          NaN        NaN  
275      NaN          NaN        NaN      NaN          NaN        NaN  
276      NaN          NaN        NaN      NaN          NaN        NaN  
277      NaN          NaN        NaN      NaN          NaN        NaN  
278      NaN          NaN        NaN      NaN          NaN        NaN  

[278 rows x 18 columns]

In [560]:
x

ResponseId.1               ProCode         ProCategory  ProRating  \
4     R_0HSeYpblz5LCLpD         Introspection       Self-Efficacy       85.0   
5     R_0HSeYpblz5LCLpD  Perspective Shifting          Creativity       90.0   
6     R_0HSeYpblz5LCLpD             Self-Care       Self-Efficacy       55.0   
10    R_0MpphsGkrwSMhuF            Creativity          Creativity       70.0   
11    R_0MpphsGkrwSMhuF   Improved Motivation       Self-Efficacy       52.0   
...                 ...                   ...                 ...        ...   
1661  R_yUcpPqMIp5DPBSN       Improved Energy     Improved Energy       80.0   
1662  R_yUcpPqMIp5DPBSN        Improved Focus      Improved Focus      100.0   
1666  R_ZfYg5JxACJX7szf           Sociability     Social Benefits       49.0   
1667  R_ZfYg5JxACJX7szf    Clarity of Thought  Cognitive Benefits       79.0   
1668  R_ZfYg5JxACJX7szf         Improved Mood       Improved Mood      100.0   

                          ConCode                ConCategory  ConRating  
4              Sleep Disturbances   Physiological Discomfort       25.0  
5                      Illegality                 Illegality        4.0  
6                          Stigma                 Illegality       20.0  
10             Sleep Disturbances   Physiological Discomfort       35.0  
11                            NaN                        NaN        NaN  
...                           ...                        ...        ...  
1661                  Preparation  Other Perceived Drawbacks        5.0  
1662                         Misc  Other Perceived Drawbacks        5.0  
1666                         None  Other Perceived Drawbacks        7.0  
1667  Unknown Risk-Effect Profile  Other Perceived Drawbacks       59.0  
1668                          NaN                        NaN        9.0  

[834 rows x 7 columns]

Now we need to create 3 sequence lists (i.e. The 4th, 5th, and 6th row per respondent and later added it to the corrrespondent column)

In [547]:
def select_last3(dataframe):
    list1 = []
    #step in range must be converted into a integer
    for i in range(3, len(x), int(len(x)/x['ResponseId.1'].nunique())):
        list1.append(i)
    return dataframe.iloc[list1]
    list2 = list1 +1


In [555]:
list1 = []
for i in range(3, len(x), int(len(x)/x['ResponseId.1'].nunique())):
        list1.append(i)
list2 = []
for j in range(4, len(x), int(len(x)/x['ResponseId.1'].nunique())):
        list1.append(j)
list3 = []
for k in range(5, len(x), int(len(x)/x['ResponseId.1'].nunique())):
        list1.append(k)
        
# or also

list2 = []
for i in list1:
    i = i+1
    list2.append(i)
list3 = []
for i in list2:
    i = i+1
    list3.append(i)

Time to assemble the three lists to the columns created.

The strategy to conduct the data analysis is the following:

- Using the shortdata for exploring most characteristics
- Using the longdata when assessing the general Pros and Cons per each user

## Merge data for final use

# Data Wrangling